Fixes #7010: Adds a linting check that ensures all TODO comments have a username with them

teddymarchildon opened the pull request:
## Fixes #7010: Adds a linting check that ensures all TODO comments have a username with them.
This adds a lint check that ensures all TODO comments have a username with them.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
Add linting function to check that TODO comments have a user

Commit:
Fix regex to match digits present in usernames

teddymarchildon commented:
I'm having a hard time adding reviewers (not quite sure why) but @seanlip @apb7 @lilithxxx PTAL whenever you get a chance! Thanks so much.

DubeySandeep commented:
Hi @teddymarchildon, I've assigned @apb7 as a reviewer for this PR! (and sorry for the trouble!)

Also, can you please check why the lint tests are failing for this PR and fix it?

teddymarchildon commented:
It looks like it is failing the linter because this PR adds a lint check to ensure that TODO comments are assigned to a user. I thought the scripts/pre_commit_linter.py only is run on the files changed in the commit, but it looks like it is run on the entire codebase in the travis-ci. So, for example, `core/domain/story_services_test` line 1358 has a TODO comment without an assigned user, but that file is not changed in this PR. 

Should I find the owner, or the last person who edited the TODO comment, of the TODO comments currently unassigned and write them in as the assigned user? It looks like the vast majority of TODO comments already have an assigned user.

DubeySandeep commented:
>  I thought the scripts/pre_commit_linter.py only is run on the files changed in the commit

Locally (as git push hook) it runs on the files changed in the feature-branch. But on travis we run these tests for the entire repo.


> Should I find the owner, or the last person who edited the TODO comment, of the TODO comments currently unassigned and write them in as the assigned user?

Yeah, this sounds good to me! Also, in case you'll find it hard to assign a username for any TODO item than you can assign it to the [code-owner](https://github.com/oppia/oppia/blob/develop/.github/CODEOWNERS) of the file. 
 

(For issues like #7010 we expect the contributor to introduce the new check as well as fix the existing issues. )

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2541,6 +2541,53 @@ def _check_npm_packages(self):
+        summary_messages = []
+        message = 'Each todo comment should be assigned to a user'
+        # Match a regular expression that is independent of the comment format.
+        todo_regex = re.compile(r'TODO\s?\(@?[A-Z|a-z|\d]+\):?\s?.+')
```
Can this be done just by adding the regex inside [BAD_PATTERNS_PYTHON_REGEXP ](https://github.com/oppia/oppia/blob/develop/scripts/pre_commit_linter.py#L267)? (not sure!)

teddymarchildon replied:
@DubeySandeep Yes! Much cleaner now. Thanks for the suggestion

Commit:
Put the bad regex for todo comments in the global variables for TS, JS, and PY files

Commit:
Add regex to HTML bad patterns, and fix TODO comments with the codeowner of the file

Commit:
Remove unnecessary comment

Commit:
Shorten line lengths in comments

Commit:
Fix conflict in the pre_commit_linter

Commit:
Add a user to a TODO comment in the jobs.py file

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into todo-comment-lint-check

Commit:
Fix TODO comment in extensions

apb7 commented on core/domain/story_services_test.py:
```diff
@@ -1355,7 +1355,7 @@ def test_record_completed_node_in_story_context(self):
 
 
-# TODO: Remove this mock class and the StoryContentsMigrationTests class
+# TODO(aks681): Remove this mock class and the StoryContentsMigrationTests class
```
Missing `@`.

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -220,6 +220,13 @@
         'excluded_dirs': ('core/tests/')
     },
+    {
+        'regexp': r'\/\/\sTODO[^\(]*[^\)]$',
```
A couple of points here:
1. I think it would be better if we put the regexp in `BAD_PATTERNS` since this needs to be checked for all files, irrespective of the type.
2. Let's enforce the `TODO(@username):` pattern instead of `TODO(username):`.

teddymarchildon replied:
Gotcha, will enforce the @ sign! 

I actually thought about putting the regexp in `BAD_PATTERNS`, but all of those patterns are checked [here](/oppia/oppia/blob/develop/scripts/pre_commit_linter.py#L2166) as just seeing if the pattern exists in the file, not with a regular expression. I would be happy to find a pattern that works for that method if you prefer! Something like `TODO: ` as a bad pattern

apb7 replied:
@teddymarchildon, I see. Can we set up a `BAD_PATTERNS_REGEXP` for this (since we have bad patterns regexp for different file types)? We can then use the same mechanism. Thanks!

apb7 replied:
Hi @teddymarchildon, have you looked into this? Thanks!

teddymarchildon replied:
Ah I must've missed that comment, working on it now!

teddymarchildon replied:
@apb7 Updated! Much cleaner - thanks for the heads up.

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -220,6 +220,13 @@
     },
+    {
+        'regexp': r'\/\/\sTODO[^\(]*[^\)]$',
+        'message': 'Please assign todo comments to a user ' +
```
`todo` --> `TODO`

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -220,6 +220,13 @@
+    {
+        'regexp': r'\/\/\sTODO[^\(]*[^\)]$',
+        'message': 'Please assign todo comments to a user ' +
+                   'in the format TODO(username) stuff todo. ',
```
`TODO(username) stuff todo` --> `TODO(@username): XXX`

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -2549,7 +2570,6 @@ def _check_npm_packages(self):
 
         return summary_messages
 
-
```
Why this change? Can you please revert this back?

apb7 reviewed (commented):
Hi @teddymarchildon, thanks for the PR. Left some comments, ptal.

Commit:
Enforce the @ sign in TODO comments over the whole project

Commit:
Fix merge conflict

Commit:
Delete files that had been deleted in develop

Commit:
Change lowercase todo to uppercase TODO in pre_commit_linter

apb7 commented:
Hi @teddymarchildon, ptal. Also, when you have addressed the review comments, please re-assign the PR  back to me for another pass. Thanks!

Commit:
Update branch with develop

Commit:
Update with develop

seanlip commented:
Sorry about this, but -- @apb7, could you explain where the `TODO(@username)` convention comes from? I always thought it was just `TODO(username)`, which is also the only version that is present in our codebase.

Commit:
Update package lock with develop

brianrodri reviewed (approved):
Code Owner LGTM

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into todo-comment-lint-check

Commit:
Checkout the package-lock file to develop branch

Commit:
Fix package-lock.json file

Commit:
Force package-lock.json to be develop version

seanlip commented:
Hi @teddymarchildon -- lint checks are failing. This is a bit odd since they're actually supposed to run locally via the pre-push hook. Could you figure out why they didn't run on your machine prior to push?

/cc @apb7 @oppia/dev-workflow-team

Commit:
Update TODO comment format in a few new locations

teddymarchildon commented:
Ah yes, it looks like after merging with develop one of these times a few new locations of TODO comments showed up without an @ sign. Should hopefully be fixed now!

I've learned that the pre_commit_linter and pre_push_linter run only on files changed in the commit locally before committing/pushing, but they are run on the whole project in the travis ci – that caused it to find some straggling examples of TODO(username) elsewhere in the project I hadn't found.

seanlip commented:
Ah, got it, thanks @teddymarchildon!

Also, per our discussion, could you please remove the "@"? I'm pretty certain that we don't need it; it's not a convention we follow.

Thanks (and sorry about the churn)!

apb7 commented:
Sorry for the confusion, @teddymarchildon! Please follow Sean's comment [here](https://github.com/oppia/oppia/pull/7114#issuecomment-512714418). Thanks!

Commit:
Merge with develop

Commit:
Revert regexp to omitting @ sign

seanlip commented on core/templates/dev/head/expressions/ExpressionSyntaxTreeService.ts:
```diff
@@ -175,7 +175,7 @@ oppia.factory('ExpressionSyntaxTreeService', [
     //   please first ensure that existing explorations do not use this
     //   parameter name. Also, to prevent future explorations using it,
-    //   modify constants.INVALID_PARAMETER_NAMES accordingly.
+    //   modify feconf.INVALID_PARAMETER_NAMES accordingly.
```
Actually I think this is in constants

teddymarchildon replied:
@seanlip Done!

seanlip commented on extensions/answer_summarizers/models_test.py:
```diff
@@ -43,8 +43,8 @@ def test_equality_of_hashable_answers(self):
 
-    # TODO(brianrodri, msl): Only non-zero answer-counts are tested. Should look
-    # into adding coverage for answers with zero-frequencies.
+    # TODO(brianrodri, msl): Only non-zero answer-counts are tested. Should
```
Drop msl from this TODO

teddymarchildon replied:
@seanlip Done!

seanlip commented on extensions/interactions/LogicProof/static/js/tools/strings.ts:
```diff
@@ -200,8 +200,9 @@ var DEFAULT_LINE_TEMPLATE_STRINGS = [{
   error: []
-  // TODO: add back or_eliminate_false template and associated logic mistakes
-  // from commit 2d9335019521 when speed issues are fixed.
+  // TODO(kevinlee12 @vojtechjelinek): add back or_eliminate_false template
```
This is TODO(Jacob)

teddymarchildon replied:
Done!

seanlip commented on extensions/interactions/MusicNotesInput/directives/OppiaInteractiveMusicNotesInputDirective.ts:
```diff
@@ -124,7 +124,8 @@ oppia.directive('oppiaInteractiveMusicNotesInput', [
 
         var NOTE_TYPE_NATURAL = 0;
-        // TODO More notes types will be added to NOTE_TYPES.
+        // TODO(kevinlee12 @vojtechjelinek)
```
TODO(wagnerdmike)

teddymarchildon replied:
Done!

seanlip reviewed (approved):
Thanks @teddymarchildon! Looks good to me, just a couple small things.

Commit:
Fix TODO assignments

DubeySandeep commented:
@seanlip, Do we need to have approval from every code-owners for this PR? (OR can we merge this directly?)

DubeySandeep commented on core/templates/dev/head/domain/classifier/AnswerClassificationResultObjectFactory.ts:
```diff
@@ -28,7 +28,7 @@ oppia.factory('AnswerClassificationResultObjectFactory', [function() {
   };
 
-  // TODO (ankita240796) Remove the bracket notation once Angular2 gets in.
+  // TODO(ankita240796) Remove the bracket notation once Angular2 gets in.
```
I'm not sure whether the pattern is `TODO(userName) <message>` OR `TODO(userName): <message>`. (or both)

seanlip replied:
Good catch! Let's go with `TODO(userName): XXX`. 

@teddymarchildon would that be possible? Sorry for not catching this earlier.

teddymarchildon replied:
Yes! Just updated the regexp to enforce the : as well as a message following the colon.

DubeySandeep commented on core/jobs.py:
```diff
@@ -1012,7 +1012,7 @@ def validate(cls, unused_mapper_spec):
             bool. Whether mapper spec and all mapper patterns are valid.
         """
-        return True  # TODO.
+        return True  # TODO(seanlip).
```
@seanlip, do we need a message here to explain the TODO?

seanlip replied:
Good catch -- yes, we do.

@teddymarchildon could this be changed to "# TODO(seanlip): Actually implement the validation.` and put on the line above `return True`?

seanlip commented:
We can merge it directly once the tests pass (which, unfortunately, they don't...).

teddymarchildon commented:
it looks like they are failing due to an HTTP Bad Gateway error when installing third party packages... not quite sure why - investigating now

DubeySandeep commented:
> it looks like they are failing due to an HTTP Bad Gateway error when installing third party packages... not quite sure why - investigating now

Initially the **ci/circleci: setup** test were also failing (because of 500 error from the docker*) but after I restarted the test twice it passed the setup tests.

Commit:
Add a message for sean's todo

kevinlee12 commented:
Github was having issues with their API earlier, which might be why.

Commit:
Put the TODO regexp in its own structure, and enforce : and message

Commit:
Fix merge conflict with develop

apb7 reviewed (approved):
LGTM! Thanks, @teddymarchildon. Nice work!

Commit:
Fix more TODO comment formats

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -147,6 +147,16 @@
+BAD_PATTERNS_REGEXP = [
+    {
+        'regexp': r'TODO[^\(]*[^\)][^:]*[^\w]*$',
+        'message': 'Please assign TODO comments to a user ' +
```
Do we need `+` here?

DubeySandeep replied:
@apb7, do we want these test to only fail for a specific pattern or for any unexpected TODO pattern?
![image](https://user-images.githubusercontent.com/16653571/61566121-ba5e3d80-aa98-11e9-8db2-0bd201055b0b.png)

teddymarchildon replied:
@DubeySandeep The + sign is used so that each line length is below 80 characters as mandated by pylint

I'm happy to enforce the uppercase starting character in this regexp, and I believe all comments have to end with a period as enforced by one of the linting processes

apb7 replied:
Hi @DubeySandeep, thanks for the comments. Please find the responses inline:

>  do we want these test to only fail for a specific pattern or for any unexpected TODO pattern?

I think the scope of this PR is to ensure that all TODO comments have an associated username in a specific pattern. Let me know if you think otherwise.


> Do we need + here?

Nice catch! No, I don't think we require `+` here. @teddymarchildon, can you please check other multi-line bad pattern messages. They do not use `+`.


> TODO comments end with a period(.) [I think we have such checks for comments]

I think the comment check will cover this as well, right?


> TODO message should start with a capital letter.

I'll file a follow-up issue for this (filed #7182). @teddymarchildon, if you're interested, please let us know on that issue thread. We'll assign it to you. Thanks!

DubeySandeep replied:
Thanks for the reply @apb7, @teddymarchildon can you please update the PR as Apurv suggested so that we can quickly merge this PR?

teddymarchildon replied:
@DubeySandeep I believe I have addressed the comments, and can start working on the next issue once this one is merged. Let me know if there is anything else to fix up or if I missed something from those comments!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2173,6 +2183,12 @@ def _check_bad_patterns(self):
+                    if _check_bad_pattern_in_file(
+                            filepath, file_content, regexp):
+                        failed = True
+                        total_error_count += 1
```
@apb7, `total_error_count` is the number of files with bad patterns? (that's also not true though) or it's the number of failed patterns?

(I mean in one file there can be multiple errors for a single bad_pattern regexp.)

apb7 replied:
I think it's the total number of failed patterns (for reference, please see other regexp checks). Thanks!

DubeySandeep reviewed (commented):
@teddymarchildon, This looks great, thanks for adding the checks and fixing the existing wrong patterns :)

I've left a few comments, PTAL.

@apb7: Can we have a follow-up issue for handling: (We can implement this latter)
- TODO comments end with a period(.) [I think we have such checks for comments]
- TODO message should start with a capital letter.

DubeySandeep commented on core/templates/dev/head/pages/get-started-page/get-started-page.mainpage.html:
```diff
@@ -46,7 +46,7 @@ <h2 translate="I18N_GET_STARTED_PAGE_PARAGRAPH_8_HEADING"></h2>
             </p>
 
-            <!-- todo: make it easy to get the right link to share, update this doc with that info -->
+            <!-- TODO(vojtechjelinek): make it easy to get the right link to share, update this doc with that info -->
```
@apb7, I think the "Period at the end of comment" doesn't work here!

DubeySandeep commented on core/tests/protractor_desktop/extensions.js:
```diff
@@ -48,7 +48,7 @@ describe('rich-text components', function() {
       richTextEditor.appendBoldText('bold');
       richTextEditor.appendPlainText(' ');
-      // TODO (Jacob) add test for image RTE component
+      // TODO(Jacob): add test for image RTE component
```
@apb7, Ditto.

DubeySandeep commented on core/tests/protractor_desktop/navigation.js:
```diff
@@ -95,11 +95,11 @@ describe('Oppia static pages tour', function() {
     general.checkForConsoleErrors([
-      // TODO (Jacob) Remove when
+      // TODO(Jacob): Remove when
       // https://code.google.com/p/google-cast-sdk/issues/detail?id=309 is fixed
```
@apb7, I see this in multiple places, ca we add this also as a part of #7182?

Commit:
Remove unnecessary + sign from message

seanlip commented:
I'm a little confused about what the next step is here. I see some comments by @DubeySandeep addressed to @apb7 and I don't know whether the ball is still in @teddymarchildon's court or not.

I'm assigning @apb7 to reply to @DubeySandeep's comments. @teddymarchildon -- would you mind replying to this thread and leaving a top-level comment if you're done with everything and would just like to get a review? I'm happy to merge this once @apb7 and @DubeySandeep confirm.

Thanks!

teddymarchildon commented:
My bad, I forgot to add a comment once I was done addressing comments. I just added one replying, but I think I have addressed the issues @DubeySandeep and @apb7 brought up - if not let me know and I can fix it up!

I can start working on the next issue once this one is merged.

seanlip commented:
Great, thanks @teddymarchildon! In that case, deassigning you and adding @DubeySandeep @apb7 for final review/confirmation.

Also, please note that you can work on more than one issue at a time! Just make sure to do that on a new, separate feature branch created from develop, as described [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).

DubeySandeep reviewed (approved):
I think the issue stated in #7010 is resolved and we can merge this PR. LGTM!

DubeySandeep commented:
@seanlip, Can we merge this PR directly (As stated [here](https://github.com/oppia/oppia/pull/7114#issuecomment-512926870))? 

Note: @apb7 has already approved this PR!

DubeySandeep commented:
@teddymarchildon, Congrats on your first successful PR to Oppia! :)

seanlip commented:
OK, merging! Congrats @teddymarchildon :D
